svc/GridAutosizeService.js

```diff
@@ -0,0 +1,182 @@
+        }
+        return ret;
+    }
+
+    //------------------
+    // Implementation
+    //------------------
+    autoSizeColumn({colId, gridModel}) {
+        try {
+            const {store} = gridModel,
+                records = [...store.records, store.summaryRecord],
+                column = gridModel.findColumn(gridModel.columns, colId);
+
+            if (!column.autoSizeOptions.enabled) return;
+
+            if (gridModel.treeMode && store.allRootCount !== store.allCount && column.isTreeColumn) {
+                // For tree columns, we need to account for the indentation at the different depths.
+                // Here we group the records by tree depth and determine the max width at each depth.
+                const recordsByDepth = groupBy(records, record => record.ancestors.length),
+                    ret = [];
+
+                forOwn(recordsByDepth, (records, depth) => {
+                    ret.push(this.calcMaxWidth(records, column, gridModel, this.getIndentation(depth)));
+                });
+
+                return max(ret);
+            } else {
+                return this.calcMaxWidth(records, column, gridModel);
+            }
+        } catch (e) {
+            console.debug(`Could not calculate width for column "${colId}".`, e);
+        } finally {
+            this.setCellElActive(false);
+        }
+    }
+
+    calcMaxWidth(records, column, gridModel, indentation = 0) {
+        const {autoSizeOptions, field, getValueFn, renderer} = column,
+            {sampleCount, buffer, minWidth, maxWidth} = autoSizeOptions,
+            useRenderer = isFunction(renderer);
+
+        // 1) Get unique values
+        const values = new Set();
+        records.forEach(record => {
+            if (!record) return;
+            const rawValue = getValueFn({record, field, column, gridModel}),
+                value = useRenderer ? renderer(rawValue, {record, column, gridModel}) : rawValue;
+            values.add(value);
+        });
+
+        // 2) Use a canvas to estimate and sort by the pixel width of the string value.
+        // Strip html tags but include parentheses / units etc. for renderers that may return html,
+        const sortedValues = sortBy(Array.from(values), value => {
+            const width = isNil(value) ? 0 : this.getStringWidth(stripTags(value.toString()));
+            return width + indentation;
+        });
+
+        // 3) Extract the sample set of longest values for rendering and sizing
+        const longestValues = sortedValues.slice(Math.max(sortedValues.length - sampleCount, 0));
+
+        // 4) Render to a hidden cell to calculate the max displayed width
+        let result = reduce(longestValues, (currMax, value) => {
+            const width = this.getCellWidth(value, useRenderer) + indentation + buffer;
+            return Math.max(currMax, width);
+        }, 0);
+
+        result = max([result, minWidth]);
+        result = min([result, maxWidth]);
```

Comment:
There is lodash `clamp` in case you would like to use that.

Reply:
In my testing, `clamp` only seems to work if *both* the lower and upper bound are defined, which we can't guarantee

Reply:
Ah yes it's weirdly (to me) asymmetrical where the lower bound is optional but the upper bound is not.